mm2src/crypto/src/global_hd_ctx.rs

```diff
@@ -92,21 +76,28 @@ impl GlobalHDAccountCtx {
     pub fn derive_secp256k1_secret(
         &self,
         derivation_path: &StandardHDPathToCoin,
+        path_to_address: &StandardHDCoinAddress,
     ) -> MmResult<Secp256k1Secret, Bip32Error> {
-        const ACCOUNT_ID: u32 = 0;
-        const CHAIN_ID: u32 = 0;
-
-        let mut account_der_path = derivation_path.to_derivation_path();
-        account_der_path.push(ChildNumber::new(ACCOUNT_ID, HARDENED).unwrap());
-        account_der_path.push(ChildNumber::new(CHAIN_ID, NON_HARDENED).unwrap());
-        account_der_path.push(self.hd_account);
-
-        let mut priv_key = self.bip39_secp_priv_key.clone();
-        for child in account_der_path {
-            priv_key = priv_key.derive_child(child)?;
-        }
+        derive_secp256k1_secret(self.bip39_secp_priv_key.clone(), derivation_path, path_to_address)
+    }
+}
 
-        let secret = *priv_key.private_key().as_ref();
-        Ok(Secp256k1Secret::from(secret))
+pub fn derive_secp256k1_secret(
+    bip39_secp_priv_key: ExtendedPrivateKey<secp256k1::SecretKey>,
+    derivation_path: &StandardHDPathToCoin,
+    path_to_address: &StandardHDCoinAddress,
+) -> MmResult<Secp256k1Secret, Bip32Error> {
+    let mut account_der_path = derivation_path.to_derivation_path();
+    account_der_path.push(ChildNumber::new(path_to_address.account, HARDENED).unwrap());
+    account_der_path.push(ChildNumber::new(path_to_address.is_change as u32, NON_HARDENED).unwrap());
+    account_der_path.push(ChildNumber::new(path_to_address.address_index, NON_HARDENED).unwrap());
+
+    let mut priv_key = bip39_secp_priv_key;
+    for child in account_der_path {
+        priv_key = priv_key.derive_child(child)?;
     }
+    drop_mutability!(priv_key);
+
+    let secret = *priv_key.private_key().as_ref();
+    Ok(Secp256k1Secret::from(secret))
```

Comment:
we shouldn't store sensitive data longer in memory than needed. could use e.g. https://crates.io/crates/zeroize or similar approach. We ult. need the key only in case of signing requirement. Will open separate issue ticket since this affects multiple layers - but figured, maybe we want to start here / with HD withdrawl layer, since critical layer?

Reply:
>  could use e.g. https://crates.io/crates/zeroize or similar approach

Thanks for the suggestion. Using `zeroize` looks like a good approach to zeroing the memory region after the variable is dropped, we should also consider disabling swapping for this memory region to avoid the OS saving secrets to disk unintentionally. Please note that the master seed is kept in memory throughout the application runtime, we should probably add the option to keep it encrypted only in memory and decrypt it using the RPC password (the rpc password is also kept in memory, so maybe we should keep the hash of the password+salt instead to verify the password on each call) as part of this separate issue.

Reply:
`drop` itself is enough for this purpose

Reply:
For certain values, we have to keep them in memory constantly to operate mm2 without always asking users. Which might not be good for GUI clients and bots. To protect against memory attacks, we can change the memory addresses dynamically while the program is running so we don't have the important values on the same address all the time while the program is running. However, we should also make sure passphrases(and other informations) are safe, as they're stored in a plain text file right know (MM_CONF_PATH). If someone gets into the mm2 device, they can quickly get the passphrases from the text file without needing doing complicated hardware attack / reverse engineering stuff.

Reply:
Yes, I know that komodod use MemoryPageLocker object for locking privkeys in memory, I guess, to prevent sending virtual pages with sensitive data to swap file. I wonder if we have the same feature in rust

Reply:
> Yes, I know that komodod use MemoryPageLocker object for locking privkeys in memory, I guess, to prevent sending virtual pages with sensitive data to swap file. I wonder if we have the same feature in rust

We can use `mlock` from libc, I don't know if there is any better way for that.

Reply:
seems this would be platform-dependant as in komodod it is implemented like
```
bool MemoryPageLocker::Lock(const void* addr, size_t len)
{
#ifdef _WIN32
    return VirtualLock(const_cast<void*>(addr), len) != 0;
#else
    return mlock(addr, len) == 0;
#endif
}
```

Reply:
Thanks for your input @dimxy @ozkanonur, I will look into it once the issue is opened and planned. We can discuss the details more then :)

Reply:
Yes, we need to use `winapi` for doing this on Windows as libc doesn't support windows. Both `winapi` and `libc` are already present in our dependency tree, so we can do it easily.